src/extension-runners/firefox-android.js

```diff
@@ -0,0 +1,653 @@
+    }
+  }
+
+  async adbDiscoveryAndForwardRDPUnixSocket() {
+    const {
+      adbClient,
+      selectedAdbDevice,
+      selectedFirefoxApk,
+      params: {
+        adb,
+        firefoxAndroidTimeout,
+      },
+    } = this;
+
+    const stdin = this.params.stdin || process.stdin;
+
+    const {
+      unixSocketDiscoveryRetryTime,
+    } = FirefoxAndroidExtensionRunner;
+
+    let {
+      unixSocketDiscoveryMaxTime,
+    } = FirefoxAndroidExtensionRunner;
+
+    if (typeof firefoxAndroidTimeout === 'number') {
+      unixSocketDiscoveryMaxTime = firefoxAndroidTimeout;
+    }
+
+    // Firefox for Android debugger socket discovery.
+    const startedDiscoveryTime = Date.now();
+    let androidUnixSockets = [];
+    let userExit = false;
+
+    const handleCtrlC = (str, key) => {
+      if (key.ctrl && key.name === 'c') {
+        userExit = true;
+      }
+    };
+
+    if (stdin.isTTY && stdin instanceof tty.ReadStream) {
+      readline.emitKeypressEvents(stdin);
+      stdin.setRawMode(true);
+
+      stdin.on('keypress', handleCtrlC);
+    }
+
+    while (!userExit && androidUnixSockets.length === 0) {
+      if (Date.now() - startedDiscoveryTime > unixSocketDiscoveryMaxTime) {
+        throw new WebExtError(
+          'Timeout while waiting for the Android Firefox Debugger Socket'
+        );
+      }
+
+      await new Promise((resolve) => {
+        setTimeout(resolve, unixSocketDiscoveryRetryTime);
+      });
+
+      log.info(`Waiting ${selectedFirefoxApk} to be ready... ` +
+               '(it can take a while on an ARM emulator)');
+      androidUnixSockets = await adbClient.shell(selectedAdbDevice.id, [
+        'cat', '/proc/net/unix',
+      ]).then(adb.util.readAll);
+
+      androidUnixSockets = androidUnixSockets.toString()
+        .split('\n').filter((line) => {
+          return line.trim().endsWith('firefox-debugger-socket');
+        });
+    }
+
+    if (stdin.isTTY && stdin instanceof tty.ReadStream) {
+      stdin.removeListener('keypress', handleCtrlC);
+    }
+
+    if (userExit) {
+      throw new UsageError(
+        'User requested exit during Android Firefox Debugger Socket discovery'
+      );
+    }
+
+    // Got a debugger socket file to connect
+    this.selectedRDPSocketFile = androidUnixSockets[0].split(/\s/)
+                                                      .slice(-2, -1)[0];
+    log.debug(`RDP Socket File selected: ${this.selectedRDPSocketFile}`);
+
+    const tcpPort = await this.chooseLocalTcpPort();
+
+    // Log the choosen tcp port at info level (useful to the user to be able
+    // to connect the Firefox DevTools to the Firefox for Android instance).
+    log.info(`Android Remote Debugging TCP Port selected: ${tcpPort}`);
+
+    await adbClient.forward(selectedAdbDevice.id, `tcp:${tcpPort}`,
+                            `localfilesystem:${this.selectedRDPSocketFile}`);
+    this.selectedTCPPort = tcpPort;
+  }
+
+  chooseLocalTcpPort(): Promise<number> {
```

Comment:
This is also something that I think we can refactor and move into a separate module, so that we can reuse in other places where it could be useful to easily choose a free tcp port.